pkg/apis/apps/kfdef/v1beta1/application_types.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	KfConfigFile = "app.yaml"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KfDef is the Schema for the applications API
+// +k8s:openapi-gen=true
+type KfDef struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   KfDefSpec   `json:"spec,omitempty"`
+	Status KfDefStatus `json:"status,omitempty"`
+}
+
+type KfDefSpec struct {
+	Applications []Application `json:"applications,omitempty"`
+	Plugins      []Plugin      `json:"plugins,omitempty"`
+	Secrets      []Secret      `json:"secrets,omitempty"`
+	Repos        []Repo        `json:"repos,omitempty"`
+}
+
+// Application defines an application to install
+type Application struct {
+	Name            string           `json:"name,omitempty"`
+	KustomizeConfig *KustomizeConfig `json:"kustomizeConfig,omitempty"`
+}
+
+type KustomizeConfig struct {
+	RepoRef    *RepoRef    `json:"repoRef,omitempty"`
+	Overlays   []string    `json:"overlays,omitempty"`
+	Parameters []NameValue `json:"parameters,omitempty"`
+}
+
+type RepoRef struct {
+	Name string `json:"name,omitempty"`
+	Path string `json:"path,omitempty"`
+}
+
+type NameValue struct {
+	Name         string `json:"name,omitempty"`
+	Value        string `json:"value,omitempty"`
+	InitRequired bool   `json:"initRequired,omitempty"`
+}
+
+// Plugin can be used to customize the generation and deployment of Kubeflow
+type Plugin struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec *runtime.Object `json:"spec,omitempty"`
```

Comment:
What would be the problem with including TypeMeta and ObjectMeta fields?

IIRC I had originally tried using RawExtension and that didn't work; I think because it doesn't define DeepCopy or possibly some other methods.

Reply:
see https://pursuit.purescript.org/packages/purescript-kubernetes/0.1.1/docs/Kubernetes.Api.Runtime

I believe this means we make it rawExtension but convert to runtime.Object when unmarshalling

Reply:
would there be a risk of using runtime.Object?  It already includes `TypeMeta` for us to identify the object, right?

Reply:
@jlewi using the Object interface means that we force the Plugins to implement the Object interface.

> IIRC I had originally tried using RawExtension and that didn't work; I think because it doesn't define DeepCopy or possibly some other methods.

Do you have more info on that.
My understanding is that RawExtension should be used for storing arbitrary JSON.
Relevant link: https://github.com/kubernetes-sigs/service-catalog/issues/911#issuecomment-305506879

@gabrielwen using runtime.Object requires plugins to implement the Object interface.

What we want to store there is arbitrary JSON and my understanding from my search is that RawExtension is what should be used.

Reply:
What's the problem with expecting plugins to implement the Object interface? Don't we need that so that we do things like generate appropriate DeepCopy methods?

We are using type Object in the v1alpha1 spec and it is working as expected. Here's the sample GCPPluginSpec

```
  plugins:
  - name: gcp
    spec:
      auth:
        iap:
          oAuthClientId: 147474701642-rn0ksbu4eikorfk8j0uvknubbke57ri1.apps.googleusercontent.com
          oAuthClientSecret:
            name: CLIENT_SECRET
      createPipelinePersistentStorage: true
```

So whether its a runtime.Object or runtime.Extension doesn't appear to influence the user facing API; seems like . So I suggest keeping it as is and if you think it should be changed open.

You can look at the existing implementation;
https://github.com/kubeflow/kubeflow/tree/master/bootstrap/pkg/kfapp/gcp 

So it doesn't look like we had to implement any special methods.

Reply:
I think to @kkasravi 's by using runtime.Object we avoid having to write our own unmarshaling logic.

Reply:
@jlewi an Object is supposed to contain TypeMeta information (https://godoc.org/k8s.io/apimachinery/pkg/runtime).

The Spec is not an Object, but part of an Object.

In other words, the Plugin struct is an Object but not the Spec field.
That's why I say it doesn't make sense to use Object for just the Spec.

This is why I propose we define Spec as RawExtension, which is also for arbitrary JSON.
Does that make sense?

Reply:
@yanniszark  Those seem like implementation concerns. How does that affect the user facing API? As the example above shows for GCP the spec would be 

```
plugins:
  - name: gcp
    spec:
      auth:
        iap:
          oAuthClientId: 147474701642-rn0ksbu4eikorfk8j0uvknubbke57ri1.apps.googleusercontent.com
          oAuthClientSecret:
            name: CLIENT_SECRET
      createPipelinePersistentStorage: true
```

From a user perspective if Object and RawExtension mean "unstructured json" then what's the difference?

If we change it to RawExtension what changes about the Object?

If the choice of RawExtension vs. Object doesn't affect the user facing API then IMO its an implementation detail. In which case we should just stick with Object for now because that's what the existing code is using so we know it works. If there's a reason that using Extension is better we can change it later. If the YAML a user writes isn't impacted then that's not a backwards incompatible change.

The docs don't make this very clear
https://godoc.org/k8s.io/apimachinery/pkg/runtime#RawExtension

> RawExtension is used to hold extensions in external versions.

> To use this, make a field which has RawExtension as its type in your external, versioned struct, and Object in your internal struct. You also need to register your various plugin types.

These docs lead me to conclude that its not really impacting the user facing API; i.e. the Json/YAML writes but rather how serialization/deserialization works.

Reply:
@gabrielwen @yanniszark It looks like I might have been confused I thought we were using Object but looks like we are currently using RawExtension.
https://github.com/kubeflow/kubeflow/blob/9455ab572f4fe287db984860b7e9452e6ac60a87/bootstrap/pkg/apis/apps/kfdef/v1alpha1/application_types.go#L107

It looks like the implementation has changed since v1alpha1.


Either way I think we want to pick the implementation that results in AP thatI looks like the following

```
kind: kfDefGcpPlugin
apiVersion: v1beta1
metadata:
  name: gcp-plugin
  namespace: kubeflow
spec:
  auth:
    iap:
      oAuthClientId: 147474701642-rn0ksbu4eikorfk8j0uvknubbke57ri1.apps.googleusercontent.com
      oAuthClientSecret:
        name: CLIENT_SECRET
  createPipelinePersistentStorage: true
```
I think the way to resolve this is to write a unittest to verify the YAML ends up looking the way we want it.

We already have such a unit test
https://github.com/kubeflow/kubeflow/blob/9455ab572f4fe287db984860b7e9452e6ac60a87/bootstrap/pkg/apis/apps/kfdef/v1alpha1/application_types_test.go#L251

So my suggestion is to port over the unittest to v1beta1 and update the YAML test file to look like what we want. Once the unittest passes we are good to go.